### PR TITLE
ngspice: update to 35 and fix issue with binary not built/installed

### LINKED ISSUE #32251: ngspice broken package
### System

* xuname:  
  * Void 5.12.19_1 x86_64 GenuineIntel notuptodate rrmFFF
* package:  
  * ngspice-33_1

### Expected behavior

ngspice package to provide the binaries and files to run spice decks.

### Actual behavior

No ngspice binary provided. Only binary provided is cmpp (Code Model PreProcessor for Xspice)

### Steps to reproduce the behavior

* install ngspice via ``xbps-install ngspice``
* Try to run spice deck.

### Additional Debug information

running ``find / 2>&1 | grep -i ngspice`` does not return anything helpful or any binaries.

Output of ``xbps-query -Rf ngspice``
```
/usr/bin/cmpp
/usr/lib/libngspice.so.0.0.0
/usr/lib/ngspice/analog.cm
/usr/lib/ngspice/digital.cm
/usr/lib/ngspice/spice2poly.cm
/usr/lib/ngspice/table.cm
/usr/lib/ngspice/xtradev.cm
/usr/lib/ngspice/xtraevt.cm
/usr/share/licenses/ngspice/COPYING
/usr/share/man/man1/cmpp.1
/usr/share/ngspice/dlmain.c
/usr/share/ngspice/scripts/ciderinit
/usr/share/ngspice/scripts/devaxis
/usr/share/ngspice/scripts/devload
/usr/share/ngspice/scripts/setplot
/usr/share/ngspice/scripts/spectrum
/usr/share/ngspice/scripts/spinit
/usr/lib/libngspice.so.0 -> /usr/lib/libngspice.so.0.0.0
```

Forcing a reinstall does not help.

Typically ngspice provides a binary in ``/usr/bin/``

### PR DESCRIPTION
**Pull request by mtboehlke**
#### General
- [ ] This is a new package and it conforms to the [quality requirements](https://github.com/void-linux/void-packages/blob/master/Manual.md#quality-requirements)

#### Have the results of the proposed changes been tested?
- [ ] I use the packages affected by the proposed changes on a regular basis and confirm this PR works for me
- [x] I generally don't use the affected packages but briefly tested this PR

close #32251

(edited PR title to be more clear)

### CODE REVIEW

**Comment by mtboehlke** [comment]
closes #32251

File: srcpkgs/ngspice/template [hunk 1]
```diff
@@ -1,28 +1,62 @@
...
 distfiles="${SOURCEFORGE_SITE}/ngspice/ng-spice-rework/${version}/${pkgname}-${version}.tar.gz"
-checksum=b99db66cc1c57c44e9af1ef6ccb1dcbc8ae1df3e35acf570af578f606f8541f1
+checksum=2263fffc6694754972af7072ef01cfe62ac790800dad651bc290bfcae79bd7b5
+
+# Configure can only handle building ngspice as a library or as a binary, but not both,
+#  so essentially we have to build ngspice twice
+pre_configure() {
+	install -d .build_bins
+	cp -fpr * .build_bins/
+}
```
Comment (ericonr) [inline srcpkgs/ngspice/template:21]: We can avoid some of this if it supports out of tree builds. So instead you'd do `../configure ...`.

Could indicate it with `configure_script`

File: srcpkgs/ngspice/template [hunk 2]
```diff
@@ -1,28 +1,62 @@
...
+
+# --with-ngshared builds the library, but not binary, so remove flag for the second pass.
+# readline should not be configured for the library, but include it here,
+#  see: https://github.com/imr/ngspice/commit/b86c85f85bbba6e45dc030af3e853edf8b9cfa3d
+post_configure() {
+	cd .build_bins
+	configure_args="--with-readline=yes $(printf %s\\n "$configure_args" | sed 's/--with-ngshared//')" do_configure
```
Comment (ericonr) [inline srcpkgs/ngspice/template:28]:
```suggestion
	configure_args="--with-readline=yes ${configure_args/--with-ngshared/}" do_configure
```

**Review by ericonr — COMMENTED** [review]
Even if unnecessary, I wonder if we should make `ngspice` depend on `libngspice`... This way people's systems won't be broken.

Also, this PR needs to revbump kicad to pick up the new dependency.

Another possible solution is creating a `ngspice-progs` subpackage, but that feels kind of ugly.

File: srcpkgs/ngspice/template (hunk 1 above)
Comment (mtboehlke) [reply srcpkgs/ngspice/template]: I am pretty unfamiliar with out of tree builds, but does look like it supports it as long as the subdir is configured first.  I reworked the template a bit but this seems to be a good solution.

File: srcpkgs/ngspice/template [hunk 3]
```diff
@@ -1,28 +1,62 @@
...
+	cd .build_bins
+	do_build
+}
 
 post_install() {
 	vlicense COPYING
+	cd .build_bins
+	do_install
+	# cmpp is not distributed in v34, so don't include man page
+	rm -f "${DESTDIR}/usr/share/man/man1/cmpp.1"
```
[the remark this answers is not in the document]
Comment (mtboehlke) [reply srcpkgs/ngspice/template]: Fixed

File: srcpkgs/ngspice/template [hunk 4]
```diff
@@ -1,28 +1,62 @@
...
+	# config.h is an internal header file, so don't include
+	#  this will probably be fixed in v35
+	rm -f "${DESTDIR}/usr/include/config.h"
```
[the remark this answers is not in the document]
Comment (mtboehlke) [reply srcpkgs/ngspice/template]: Fixed

**Comment by mtboehlke** [comment]
I included the revbump to kicad.

What exactly would happen if we don't have ngspice depend on libngspice?

**Comment by mtboehlke** [comment]
I see what you're saying.

As far as the checks go, apparently since v34 you have to add a .spiceinit file to the home directory containing `set ngbehavior=mc`.  They pass on my machine after doing that.

**Comment by ericonr** [comment]
You can create a `pre_check` exporting a custom `HOME` and creating the file, if you want to run the tests :)

**Comment by mtboehlke** [comment]
Well, while playing with this I noticed that version 35 has been released. Removing the erroneously included files is no longer needed.  Thanks for the tip for including pre_check!  The only thing I haven't included is a libngspice dependency in ngspice, since it doesn't really seem necessary and doubles the install size for those just looking to use ngspice (admittedly install size probably isn't a big deal either).

File: srcpkgs/ngspice/template [hunk 5]
```diff
@@ -1,28 +1,60 @@
...
+# library built during do_build, now build the binary
+post_build() {
+	unset make_build_args
+	do_build
+}
+
+pre_check() {
+	export HOME="$(pwd)/.home_dir"
+	install -d "$HOME"
+	printf 'set ngbehavior=mc\n' > "${HOME}/.spiceinit"
```
Comment (ericonr) [inline srcpkgs/ngspice/template:36]: Maybe have a comment here as to why?

Comment (mtboehlke) [reply srcpkgs/ngspice/template:38]: Good call.

**Comment by mtboehlke** [comment]
:)

File: srcpkgs/ngspice/template [hunk 6]
```diff
@@ -1,28 +1,62 @@
...
+pre_configure() {
+	install -d .build_lib
+	cd .build_lib
+	../configure --with-ngshared ${configure_args/--with-readline=yes/}
+}
+
+# library built during first do_build, now build the binary
+post_build() {
+	make_build_args= do_build
+}
```
Comment (sgn) [inline srcpkgs/ngspice/template:30]: I think it's better to remove `--with-readline=yes` from `configure_args` then
```suggestion
do_configure() {
	mkdir build-lib
	cd build-lib
	../configure --with-ngshared ${configure_args}
	mkdir build
	../configure --with-readline=yes ${configure_args}
}

do_build() {
	make -C build-lib ${makejobs}
	make -C build ${makejobs}
}
```

File: srcpkgs/ngspice/template [hunk 7]
```diff
@@ -1,28 +1,62 @@
...
+	export HOME="${PWD}/.home_dir"
+	install -d "$HOME"
+	printf 'set ngbehavior=mc\n' > "${HOME}/.spiceinit"
+}
 
 post_install() {
 	vlicense COPYING
+	cd .build_lib
+	do_install
 }
```
Comment (sgn) [inline srcpkgs/ngspice/template:45]: And:
```suggestion
do_install() {
	make -C build-lib install
	make -C build install
	vlicense COPYING
}
```

Comment (mtboehlke) [reply srcpkgs/ngspice/template:45]: It does seem clearer this way.  I kind of liked not touching do_install, etc., to keep defaults and things, but maybe that's just obscuring things.`

File: srcpkgs/ngspice/template (hunk 6 above)
Comment (mtboehlke) [reply srcpkgs/ngspice/template]: Applied suggestion, with just minor tweaks

**Comment by mtboehlke** [comment]
I am pretty happy with this version of the template.  Functionally these last set of changes to the same as the last, but make it clearer what it's doing.  Thanks for all the helpful feedback so far!

**Comment by mtboehlke** [comment]
I went ahead and did a rebase since my branch was old.

**Comment by ericonr** [comment]
> My question is how best to handle that. Should they be moved to libngspice and have ngspice then depend on it as well, make another subpackage called something like ngspice-common, or maybe not split off libngspice at all?

I'm honestly ok with this option.

**Comment by mtboehlke** [comment]
Sounds good!  I went ahead and pushed changes to no longer split a libngspice package.  Builds okay on my system.

**Comment by ericonr** [comment]
Ok, I think this still requires a kicad rebuild for some godforsaken reason :/

I can test and confirm tomorrow.

**Comment by mtboehlke** [comment]
Oh, there might be a issue if the version in the repo was linked against readline, which I'm pretty sure it is.  We may want to rebuild kicad anyway just to be safe.

**Comment by ericonr** [comment]
It's worse ;) it tries to dlopen `libngspice.so.0.0.0`, but the new file is `libngspice.so.0.0.1`

**Comment by ericonr** [comment]
Thanks! I pushed the kicad change and merged.

**Comment by mtboehlke** [comment]
:)  Thanks ericonr!

**Pull request merged**